template/python3-http/index.py

```diff
@@ -2,11 +2,20 @@
 from flask import Flask, request, jsonify
 from waitress import serve
 import os
+import sys
+
+import signal
 
 from function import handler
 
 app = Flask(__name__)
 
+def SignalHandler(SignalNumber, Frame):
+    timeout = os.getenv("write_timeout")
+
+    sys.stderr.write('Function got SIGTERM, draining for up to: {}\n'.format(timeout))
+    sys.stderr.flush()
```

Comment:
Am i missing something? i don't see how `timeout` is used here to pause for draining. Is it just that sigterm is now completely ignored by python and it will wait until watchdog does an explicit kill?

Reply:
Exactly, you're spot on :+1:

Reply:
For context - there are lots of closed PRs trying to add graceful shutdown to Waitress - none have been accepted:

https://github.com/Pylons/waitress/issues/

In any case, whether there is a graceful shutdown of Waitress or not - the watchdog will still forcibly terminate the process at the maximum timeout.

On the plus side - if we were able to call close on the Waitress server would cause the whole program to exit earlier.

I'm not sure if this note is helpful or an accepted solution? https://github.com/Pylons/waitress/issues/198#issuecomment-1012929837

Reply:
I ran a quick test according to the comment above, it caused the server to exit immediately without waiting for existing connections to drain.

There's no documentation in the Waitress code about the close method on the server type:
 
https://github.com/Pylons/waitress/blob/455f2a5fd38cafab95893ca170c7504e4239d985/src/waitress/server.py#L357

Reply:
Given the dozen or so closed / ignored PRs for graceful shutdown, it doesn't seem like the maintainer wants to fix the issue with Waitress.

Gunicorn may provide a way to shutdown gracefully: https://docs.gunicorn.org/en/stable/signals.html?highlight=graceful%20shutdown

Waitress was chosen for being pure Python, so Gunicorn may cause issues when running on Arm, where a pre-compiled wheel may not be available.

We'd have to test this if we wanted to consider migrating.

Reply:
perhaps it would be good to document this in the code, to avoid any future confusion

Reply:
What specifically should we document?

About waitress as a choice or that Waitress lacks a graceful shutdown option?

Reply:
@LucasRoesler This is on the agenda for the community call - will you be joining us this week?